pkg/rtc/participant.go

```diff
@@ -666,50 +668,84 @@ func (p *ParticipantImpl) GetAudioLevel() (level uint8, active bool) {
 	return
 }
 
-func (p *ParticipantImpl) GetConnectionQuality() livekit.ConnectionQuality {
+func (p *ParticipantImpl) GetConnectionQuality() *livekit.ConnectionQualityInfo {
 	// avg loss across all tracks, weigh published the same as subscribed
 	var pubLoss, subLoss uint32
 	var reducedQualityPub bool
 	var reducedQualitySub bool
+	var audioScores float64
+	var numAudioTracks int
 	p.lock.RLock()
 	defer p.lock.RUnlock()
 	for _, pubTrack := range p.publishedTracks {
 		if pubTrack.IsMuted() {
 			continue
 		}
-		pubLoss += pubTrack.PublishLossPercentage()
-		publishing, registered := pubTrack.NumUpTracks()
-		if registered > 0 && publishing != registered {
-			reducedQualityPub = true
+		// audio has scores available
+		if pubTrack.Kind() == livekit.TrackType_AUDIO {
+			audioScores += pubTrack.GetConnectionScore()
+			numAudioTracks++
+		} else {
+			pubLoss += pubTrack.PublishLossPercentage()
+			publishing, registered := pubTrack.NumUpTracks()
+			if registered > 0 && publishing != registered {
+				reducedQualityPub = true
+			}
 		}
 	}
-	numTracks := uint32(len(p.publishedTracks))
-	if numTracks > 0 {
-		pubLoss /= numTracks
-	}
 
 	for _, subTrack := range p.subscribedTracks {
 		if subTrack.IsMuted() {
 			continue
 		}
-		if subTrack.DownTrack().GetForwardingStatus() != sfu.ForwardingStatusOptimal {
-			reducedQualitySub = true
+		if subTrack.DownTrack().Kind() == webrtc.RTPCodecTypeAudio {
+			audioScores += subTrack.DownTrack().GetConnectionScore()
+			numAudioTracks++
+		} else {
+			if subTrack.DownTrack().GetForwardingStatus() != sfu.ForwardingStatusOptimal {
+				reducedQualitySub = true
+			}
+			subLoss += subTrack.SubscribeLossPercentage()
 		}
-		subLoss += subTrack.SubscribeLossPercentage()
 	}
-	numTracks = uint32(len(p.subscribedTracks))
+
+	avgScore := 0.0
+	// find avg loss and avg scores across pub/sub
+	// audio scores fall are in range of [0 - 5] . 5 being best . find avg
+	if numAudioTracks > 0 {
+		avgScore = (audioScores) / float64(numAudioTracks)
+	}
+
+	numTracks := len(p.publishedTracks) + len(p.subscribedTracks) - numAudioTracks
+	avgLoss := 0.0
+	// non audio scores fall in range [x - 0] . 0 being best . find average
 	if numTracks > 0 {
-		subLoss /= numTracks
+		avgLoss = float64(pubLoss+subLoss) / float64(numTracks)
 	}
 
-	avgLoss := (pubLoss + subLoss) / 2
+	// map video to avgLossScores (> 4 -> bad[2.0], (4><2 -> good[3.5]), (<2 -> excellent[4.4])
+	avgLossScore := 3.5
 	if avgLoss >= 4 {
-		return livekit.ConnectionQuality_POOR
+		avgLossScore = 2.0
 	} else if avgLoss <= 2 && !reducedQualityPub && !reducedQualitySub {
```

Comment:
would there be any scenario where it'll be a 5? also, if there are no video tracks published/subscribed, we should just use the audio score and ignore video score. I think currently it'll still give it a avgLossScore of 4.5?

Reply:
Fixed it. When there is no loss and !reducedQuality, score will be 5